### PR TITLE
test: add wait_for_cookie_credentials() to framework for rpcwait tests

### LINKED ISSUE #18684: interface_bitcoin_cli.py fails intermittently
https://travis-ci.org/github/bitcoin/bitcoin/jobs/676156661#L6349

```
 test  2020-04-17T14:09:31.865000Z TestFramework (ERROR): 
  Called Process failed with 'error: Could not locate RPC credentials. No authentication cookie could be found, and RPC password is not set.  See -rpcpassword and -stdinrpcpass.  Configuration file: (/home/travis/build/bitcoin/bitcoin/ci/scratch/test_runner/test_runner_₿_🏃_20200417_132645/interface_bitcoin_cli_109/node0/bitcoin.conf) 
```

cc @jonatack

### PR DESCRIPTION
**Pull request by jonatack**
This PR adds a `wait_for_cookie_credentials()` method to the test framework and calls it before the `-rpcwait` tests, to avoid an intermittent race condition on the CI run with Valgrind where the cookie file isn't written yet when the CLI call with `-rpcwait` arrives to `get_auth_cookie()`.

To reproduce/test, build with
```diff
diff --git a/src/httprpc.cpp b/src/httprpc.cpp
index 60c4d06f12..3dd06c4758 100644
--- a/src/httprpc.cpp
+++ b/src/httprpc.cpp
@@ -291,6 +291,7 @@ static bool InitRPCAuthentication()
 bool StartHTTPRPC()
 {
     LogPrint(BCLog::RPC, "Starting HTTP RPC server\n");
+    UninterruptibleSleep(std::chrono::seconds{11});
     if (!InitRPCAuthentication())
```
then run the test normally and with valgrind
```
test/functional/interface_bitcoin_cli.py -l debug
valgrind test/functional/interface_bitcoin_cli.py -l debug
```
Thanks to Marco Falke for all the help. Closes #18684.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
The error was "  Called Process failed with 'error: Could not locate RPC credentials. No authentication cookie could be found, and RPC password is not set.  See -rpcpassword and -stdinrpcpass.  Configuration file: (/home/travis/build/bitcoin/bitcoin/ci/scratch/test_runner/test_runner_₿_🏃_20200417_132645/interface_bitcoin_cli_109/node0/bitcoin.conf) "

**Comment by jonatack** [comment]
Yes. I'm still looking at it, but since valgrind runs don't work out-of-repo, I wanted to see if the valgrind CI passes with this change while looking further.

**Comment by jonatack** [comment]
Am noting that no other tests use `get_auth_cookie`... seems undertested...

**Comment by MarcoFalke** [comment]
`get_auth_cookie` is used by every single test. It reads the auth cookie from disk, though it fails when the file was not yet created.

**Comment by MarcoFalke** [comment]
Btw, this happens on cirrus reliably: https://cirrus-ci.com/task/6597545033990144?command=functional_test#L1053

**Comment by MarcoFalke** [comment]
Or, if you want to test locally:

```diff
diff --git a/src/httprpc.cpp b/src/httprpc.cpp
index 60c4d06f12..3dd06c4758 100644
--- a/src/httprpc.cpp
+++ b/src/httprpc.cpp
@@ -291,6 +291,7 @@ static bool InitRPCAuthentication()
 bool StartHTTPRPC()
 {
     LogPrint(BCLog::RPC, "Starting HTTP RPC server\n");
+    UninterruptibleSleep(std::chrono::seconds{11});
     if (!InitRPCAuthentication())
         return false;

**Comment by jonatack** [comment]
Thanks! Trying that now.

**Comment by jonatack** [comment]
Thanks @MarcoFalke for your suggestions! Being able to reproduce makes all the difference. I updated the PR title and description and pushed a fix that seems to work.

**Comment by jonatack** [comment]
I see that the CI for the latest merge on master (bump wallet_import timeout) failed for this same reason: https://travis-ci.org/github/bitcoin/bitcoin/jobs/676639823

**Review by maflcko — APPROVED** [review]

File: test/functional/test_framework/util.py [hunk 1]
```diff
@@ -354,7 +354,7 @@ def get_auth_cookie(datadir, chain):
             password = split_userpass[1]
     except OSError:
         pass
-    if user is None or password is None:
+    if (user is None or password is None) and wait_for_cookie is False:
```
[the remark this answers is not in the document]
Comment (jonatack) [reply test/functional/test_framework/util.py:357]: Thanks for reviewing and your help. If I understand your comment correctly, I modified the conditional here not to change the interface, but to not raise when we're waiting for the cookie auth credentials. When they do become available, it still returns the auth pair which we're waiting for in `wait_for_cookie_credentials()`.

[the remark this answers is not in the document]
Comment (jonatack) [reply test/functional/test_framework/util.py:357]: E.g. since `wait_for_cookie` is false by default, there is no change in behavior except when it's called by `wait_for_cookie_credentials`.

[the remark this answers is not in the document]
Comment (MarcoFalke) [reply test/functional/test_framework/util.py:357]: Yeah, I don't understand why `wait_for_cookie` is needed? Just because it is turned off by default doesn't mean that it doesn't complicate the interface

[the remark this answers is not in the document]
Comment (jonatack) [reply test/functional/test_framework/util.py:357]: Hm, maybe it isn't. I modified that line before changing the conditional on line 340, because I was seeing empty creds. Good question. Re-checking.

[the remark this answers is not in the document]
Comment (MarcoFalke) [reply test/functional/test_framework/util.py:357]: They should never be empty. 

Either both are a string of length of at least 1 or a ValueError is thrown

[the remark this answers is not in the document]
Comment (jonatack) [reply test/functional/test_framework/util.py:357]: It seems needed... I'll look more into why tomorrow, since it's getting late. Would you prefer a separate function to not change `get_auth_cookie`?

[the remark this answers is not in the document]
Comment (jonatack) [reply test/functional/test_framework/util.py:357]: I was seeing empty values, which is why I had to check for them at https://github.com/bitcoin/bitcoin/pull/18691/files#diff-86294e5ae5283eebdd9f98d79007a0e1R257

[the remark this answers is not in the document]
Comment (jonatack) [reply test/functional/test_framework/util.py:357]: I'll look further into what and why.

[the remark this answers is not in the document]
Comment (jonatack) [reply test/functional/test_framework/util.py:357]: Ok, I think I see what you mean.

[the remark this answers is not in the document]
Comment (jonatack) [reply test/functional/test_framework/util.py:357]: Good catch, fixed.

**Comment by jonatack** [comment]
Simplified.

File: test/functional/test_framework/test_node.py [hunk 2]
```diff
@@ -243,6 +244,20 @@ def wait_for_rpc_connection(self):
...
+        poll_per_s = 4
+        for _ in range(poll_per_s * self.rpc_timeout):
+            try:
+                user, passphrase = get_auth_cookie(self.datadir, self.chain)
+                self.log.debug("Cookie credentials retrieved: user={}, passphrase={}".format(user, passphrase))
+                return
+            except ValueError:
+                pass
+            time.sleep(1.0 / poll_per_s)
+        self._raise_assertion_error("Unable to retrieve cookie credentials")
```
Comment (MarcoFalke) [inline test/functional/test_framework/test_node.py:259]: nit: Could rebase and adjust the error to mention the timeout (like the one above, which you can see after rebase)

File: test/functional/test_framework/test_node.py [hunk 3]
```diff
@@ -243,6 +244,20 @@ def wait_for_rpc_connection(self):
...
         self._raise_assertion_error("Unable to connect to bitcoind")
 
+    def wait_for_cookie_credentials(self):
+        self.log.debug("Waiting for cookie credentials")
+        # Poll at a rate of four times per second
+        poll_per_s = 4
+        for _ in range(poll_per_s * self.rpc_timeout):
+            try:
+                user, passphrase = get_auth_cookie(self.datadir, self.chain)
+                self.log.debug("Cookie credentials retrieved: user={}, passphrase={}".format(user, passphrase))
```
Comment (MarcoFalke) [inline test/functional/test_framework/test_node.py:254]: nit: the values aren't needed, only their presence, so no need to log them?

File: test/functional/test_framework/test_node.py [hunk 4]
```diff
@@ -243,6 +244,20 @@ def wait_for_rpc_connection(self):
...
+                return
+            except ValueError:
```
Comment (MarcoFalke) [inline test/functional/test_framework/test_node.py:256]: nit: In a future cleanup the three occurrences of `ValueError` can be changed to a self-made error `RpcMissingCred` (or so). Not asking you to do this here.

**Review by maflcko — APPROVED** [review]
ACK 83b037921fac34d4c85ce6a2e137407596fed91e

Happy to merge with or without feedback addressed. Just let me know what you prefer.

File: test/functional/test_framework/test_node.py (hunk 3 above)
Comment (jonatack) [reply test/functional/test_framework/test_node.py]: done

File: test/functional/test_framework/test_node.py (hunk 2 above)
Comment (jonatack) [reply test/functional/test_framework/test_node.py]: good idea; done

File: test/functional/test_framework/test_node.py (hunk 4 above)
Comment (jonatack) [reply test/functional/test_framework/test_node.py]: Good idea; will look for a follow-up

Comment (jonatack) [reply test/functional/test_framework/test_node.py]: > nit: In a future cleanup the three occurrences of `ValueError` can be changed to a self-made error `RpcMissingCred` (or so). Not asking you to do this here.

Good idea, was also just noticing the three occurrences of "No RPC credentials", etc. Will look at a follow-up.

**Comment by jonatack** [comment]
Took the suggestions, and also updated two cli calls in the test so the debug log shows:
```diff
- TestFramework.bitcoincli (DEBUG): Running bitcoin-cli command: None
+ TestFramework.bitcoincli (DEBUG): Running bitcoin-cli command: -getinfo

- TestFramework.bitcoincli (DEBUG): Running bitcoin-cli command: None
+ TestFramework.bitcoincli (DEBUG): Running bitcoin-cli command: -version
```

**Review by maflcko — APPROVED** [review]

**Comment by jonatack** [comment]
Rebased and added a docstring to `wait_for_cookie_credentials()`.

File: test/functional/interface_bitcoin_cli.py [hunk 5]
```diff
@@ -76,20 +75,17 @@ def run_test(self):
         else:
             self.log.info("*** Wallet not compiled; cli getwalletinfo and -getinfo wallet tests skipped")
 
-        self.stop_node(0)
-
         self.log.info("Test -version with node stopped")
-        cli_response = self.nodes[0].cli("-version").send_cli()
+        self.stop_node(0)
+        cli_response = self.nodes[0].cli().send_cli('-version')
```
Comment (MarcoFalke) [inline test/functional/interface_bitcoin_cli.py:80]: Why is this changed?

Comment (MarcoFalke) [reply test/functional/interface_bitcoin_cli.py:80]: -version is not an rpc command, but a command line option

File: test/functional/interface_bitcoin_cli.py [hunk 6]
```diff
@@ -52,7 +51,7 @@ def run_test(self):
         self.log.info("Test -getinfo returns expected network and blockchain info")
         if self.is_wallet_compiled():
             self.nodes[0].encryptwallet(password)
-        cli_get_info = self.nodes[0].cli('-getinfo').send_cli()
+        cli_get_info = self.nodes[0].cli().send_cli('-getinfo')
```
Comment (MarcoFalke) [inline test/functional/interface_bitcoin_cli.py:54]: same here

Comment (jonatack) [reply test/functional/interface_bitcoin_cli.py:54]: Changed to improve the debug log output:
```diff
- TestFramework.bitcoincli (DEBUG): Running bitcoin-cli command: None
+ TestFramework.bitcoincli (DEBUG): Running bitcoin-cli command: -getinfo

- TestFramework.bitcoincli (DEBUG): Running bitcoin-cli command: None
+ TestFramework.bitcoincli (DEBUG): Running bitcoin-cli command: -version
```
Can revert in #18594 that also touches the testfile.

Comment (jonatack) [reply test/functional/interface_bitcoin_cli.py:54]: I changed it to improve the debug log output:
```diff
- TestFramework.bitcoincli (DEBUG): Running bitcoin-cli command: None
+ TestFramework.bitcoincli (DEBUG): Running bitcoin-cli command: -getinfo

- TestFramework.bitcoincli (DEBUG): Running bitcoin-cli command: None
+ TestFramework.bitcoincli (DEBUG): Running bitcoin-cli command: -version
```
I can revert it in #18594 that also touches the testfile or look at improving the logging output for these cases.

Comment (MarcoFalke) [reply test/functional/interface_bitcoin_cli.py:54]: That should be fixed by logging the options, not passing the command line option as the rpc command

Comment (jonatack) [reply test/functional/interface_bitcoin_cli.py:54]: > That should be fixed by logging the options, not passing the command line option as the rpc command

Done in #18712.
